File: one-file.py
import os
import fastapi
import aiohttp

from dotenv import load_dotenv

load_dotenv()
app = fastapi.FastAPI()

async def client_request(request: dict) -> dict:
    """Make a request to the target API."""

    #? Start the request
    async with aiohttp.ClientSession(timeout=aiohttp.ClientTimeout(connect=5, total=500.0)) as session:
        async with session.request(
            method=request.get('method', 'POST'),
            url=request['url'],
            json=request.get('payload'),
            headers=request.get('headers', {})
        ) as response:
            response.raise_for_status()

            async for response_chunk in response.content.iter_any():
                yield response_chunk

@app.route('/v1/{path:path}', methods=['GET', 'POST', 'PUT', 'DELETE', 'PATCH', 'OPTIONS'])
async def handle(request: fastapi.Request):
    path = request.url.path.replace('/v1/v1', '/v1') # fix double v1

    try: payload = await request.json()
    except: payload = {}

    request = {
        'path': path,
        'method': request.method,
        'payload': payload,
        'headers': {
            'Content-Type': 'application/json',
            'Authorization': 'Bearer ' + os.environ['OPENAI_API_KEY']
        },
        'url': 'https://api.openai.com' + path
    }

    return fastapi.responses.StreamingResponse(client_request(request))

if __name__ == '__main__':
    import uvicorn
    uvicorn.run(app, host='0.0.0.0', port=4242)

File: one-file.go
package main

import (
	"fmt"
	"io"
	"net/http"
	"os"

	"github.com/gin-gonic/gin"
)

func main() {
	r := gin.Default()

	r.Any("/v1/*path", func(c *gin.Context) {
		path := c.Param("path")
		targetURL := "https://api.openai.com" + path

		client := &http.Client{}
		req, err := http.NewRequest(c.Request.Method, targetURL, c.Request.Body)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}

		req.Header.Set("Content-Type", "application/json")
		req.Header.Set("Authorization", "Bearer "+os.Getenv("OPENAI_API_KEY"))

		resp, err := client.Do(req)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}
		defer resp.Body.Close()

		for key, values := range resp.Header {
			for _, value := range values {
				c.Writer.Header().Set(key, value)
			}
		}

		c.Status(resp.StatusCode)

		_, err = io.Copy(c.Writer, resp.Body)
		if err != nil {
			fmt.Println(err)
		}
	})

	r.Run(":4242")
}
